### ISSUE TITLE
Missing AbstractToken attribute

### ISSUE THREAD

**Issue by coreyjadams** [mpi4jax/mpi4jax#232]
Hi all,

Last week, we upgraded our nvidia A100 system to a new programming environment, and it's been a bumpy ride.  I'm trying to install mpi4jax and test on our development system.  Every which way I try, I seem to hit this issue at runtime:

```python
  File "/soft/applications/jax/0.4.26/lib/python3.11/site-packages/mpi4jax/_src/collective_ops/bcast.py", line 76, in bcast
    res, token = mpi_bcast_p.bind(x, token, root=root, comm=comm)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/soft/applications/jax/0.4.26/lib/python3.11/site-packages/jax/_src/core.py", line 422, in bind
    return self.bind_with_trace(find_top_trace(args), args, params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/soft/applications/jax/0.4.26/lib/python3.11/site-packages/jax/_src/core.py", line 425, in bind_with_trace
    out = trace.process_primitive(self, map(trace.full_raise, args), params)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/soft/applications/jax/0.4.26/lib/python3.11/site-packages/jax/_src/core.py", line 913, in process_primitive
    return primitive.impl(*tracers, **params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/soft/applications/jax/0.4.26/lib/python3.11/site-packages/jax/_src/dispatch.py", line 87, in apply_primitive
    outs = fun(*args)
           ^^^^^^^^^^
AttributeError: 'AbstractToken' object has no attribute 'ndim'
```

Have you seen this issue?  I can't reproduce it until I actually try to do an mpi Op with jax (the first line above is a broadcast of random number keys, for example).

Any debugging suggestions would be appreciated, thanks!

[Edit]
By the way, this is - 
OS: SLES15
CUDA: 12.3
CUDNN: 8.9.7
Python: 3.11
GCC: 12 *
JAX: 0.4.26
mpi4jax: 0.3.10 tag

* GCC has some funny business that I'm still working through, occasionally the cuda compiler will pickup gcc7 if you're not careful.  I don't think that is responsible for this AttributeError.

**Comment by coreyjadams** [comment]
The line that triggers this issue is here:

```python
master_key, token = mpi4jax.bcast(master_key, root=0, comm=MPI.COMM_WORLD)
```

In this case, the type of `master_key` is an array technically, it's a rng key.  For instance, if I add some debugging to the bcast function in mpi4jax:

```python
    print("Type(x): ", type(x))

    print("Token: ", token, flush=True)
    print("type(Token): ", type(token), flush=True)
    print(token.ndim)
```
I get this:



```
Type(x):  <class 'jaxlib.xla_extension.ArrayImpl'>
Token:  <jax._src.core.Token object at 0x14ad33dc1010>
type(Token):  <class 'jax._src.core.Token'>
```

I found this in the changelog for jax 0.4.27:

```core.Token now is a non-trivial class which wraps a jax.Array. It could be created and threaded in and out of computations to build up dependency. The singleton object core.token has been removed, users now should create and use fresh core.Token objects instead.
```

Which makes me think this version of jax is looking to use that token implementation but I've built JAX 0.4.26.    It's really strange ...

I will try to get the development branch of JAX and see if that fixes things.

**Comment by dionhaefner** [comment]
Wait, are you really using `mpi4jax==0.3.10`? That's seriously outdated, and [only compatible with JAX up to 0.3.16](https://github.com/mpi4jax/mpi4jax/blob/v0.3.10/mpi4jax/_src/_latest_jax_version.txt).

**Comment by coreyjadams** [comment]
Ah no sorry that was me trying every which way I could to make the versions agree with each other.  I'm reproducing this with the latest and greatest development branch.  JAX @0.4.26.

```python
>>> import mpi4jax
mpi4>>> mpi4jax.__version__
'0.4.1+0.gffd431b.dirty'
>>> 
```

**Comment by dionhaefner** [comment]
I can't reproduce this, and CI is still passing. Could you please create a self-contained script that reproduces the error?

**Comment by coreyjadams** [comment]
It took me some tweaking to find a way to reproduce this standalone.  Here's what I've got, and I did a fresh install of JAX to (hopefully) make things reproducible:

CUDA: 12.3.2
CUDNN: 8.9.7
JAX: 0.4.26 via `pip install "jax[cuda12_local]" ...
mpi4jax: 0.4.1
mpi4py: 3.1.5
MPI: mpich, 1.3.4

```python
import sys, os
import jax
import jax.numpy as numpy
import jax.random as random

import time

# Turn this line on or off to for JAX and XLA onto single devices or not
os.environ["CUDA_VISIBLE_DEVICES"]= os.environ["PALS_LOCAL_RANKID"]

from mpi4py import MPI
import mpi4jax


rank = MPI.COMM_WORLD.Get_rank()
size = MPI.COMM_WORLD.Get_size()

# Select the device, the logic here is controlled by the number of visible devices:
devices = jax.local_devices()
if len(devices) == 1:
  device = devices[0]
else:
  device = jax.local_devices()[rank]

with jax.default_device(device):
    key = random.PRNGKey(0)
    # Committing the array to the device appears to be critical to triggering the token.ndim error
    key = jax.device_put(key, device)

# Can check what device the array is on, only useful if CUDA_VISIBLE_DEVICES is unset:
print(f"Rank {rank} as data on devices {key.devices()}")

# Make it not all zeros:
key = random.split(key)

if rank == 0: 
  # Change just the key that will get broadcast to check that it's changing:
  key = random.split(key[0])
print(f"{rank} Pre key: {key}")



key, token = mpi4jax.bcast(key, 0)

print(f"{rank} Post key: {key}")
```

I ran like this, nothing fancy: `mpiexec -n 4 -ppn 4 python mpi4jax_test_bcast.py`, on a node with 4x A100.

I found there are two lines in this script that could affect the outcome: Whether to mask devices via CUDA, and whether to use device_put() to "commit" the data to a device.

- When _not using_ CUDA_VISIBLE_DEVICES (line commented 9 out) and _not_ using `device_put` (line 28), the script fails with allocation issues.  I believe JAX is allocating to GPU0 here from all 4 ranks.
- When _using_  CUDA_VISIBLE_DEVICES (line commented 9 out) and _not_ using `device_put` (line 28), the script appears to run correctly.  This seems to rely on making sure JAX can only see one device, and that jax will allocate to device 0.  But, I've seen instances where setting `CUDA_VISIBLE_DEVICES` after mpi_init goes poorly: `CUDA_VISIBLE_DEVICES` does not always get respected after MPI_Init. (In this case, on this OS and MPI, it did, but it's been a pain point in the past)
- Whether or not using CUDA_VISIBLE_DEVICES (line commented 9 out) and _with_ using `device_put` (line 28), the script runs to the bcast line (L43) but fails with `AbstractToken' object has no attribute `ndim`.


So, putting things together, it seems to trigger this missing AttributeError only when using `jax.device_put`.

Hopefully, this will let you reproduce it, otherwise I am just going crazy :)

**Comment by dionhaefner** [comment]
Thanks, I can reproduce it now. It seems like specifying a device triggers some sharding logic within JAX which can't handle tokens. I'll raise an issue with the JAX devs.

As a workaround, you can try our experimental `notoken` mode to use primitives with automatic token handling:

```python
import sys, os
import jax
import jax.numpy as numpy
import jax.random as random

import time

# Turn this line on or off to for JAX and XLA onto single devices or not
os.environ["CUDA_VISIBLE_DEVICES"]= os.environ["PALS_LOCAL_RANKID"]

from mpi4py import MPI
from mpi4jax.experimental.notoken import bcast


rank = MPI.COMM_WORLD.Get_rank()
size = MPI.COMM_WORLD.Get_size()

# Select the device, the logic here is controlled by the number of visible devices:
devices = jax.local_devices()
if len(devices) == 1:
  device = devices[0]
else:
  device = jax.local_devices()[rank]

with jax.default_device(device):
    key = random.PRNGKey(0)
    # Committing the array to the device appears to be critical to triggering the token.ndim error
    key = jax.device_put(key, device)

# Can check what device the array is on, only useful if CUDA_VISIBLE_DEVICES is unset:
print(f"Rank {rank} as data on devices {key.devices()}")

# Make it not all zeros:
key = random.split(key)

if rank == 0: 
  # Change just the key that will get broadcast to check that it's changing:
  key = random.split(key[0])
print(f"{rank} Pre key: {key}")



key = bcast(key, 0)

print(f"{rank} Post key: {key}")
```

Alternatively, you can downgrade JAX to 0.4.24 where the original script worked for me.

**Comment by coreyjadams** [comment]
Will `notoken` respect ordering of ops even in JIT compiled code?  If so, that's a workable path forward and I'll try it.

Thank you for looking at this!

**Comment by dionhaefner** [comment]
> Will notoken respect ordering of ops even in JIT compiled code?

Yes, if you apply `notoken` to *all* primitives. Otherwise you might have to use the envvar `MPI4JAX_PREFER_NOTOKEN=1` to ensure all ops use the new effects, even when you're still passing a token.

**Comment by dionhaefner** [comment]
Raised here: https://github.com/google/jax/issues/21160

**Comment by coreyjadams** [comment]
I had to do a fancy song and dance with the compilers to get things to build.  I confirm that the fix for google/jax#21160 fixed this issue for me in mpi4jax in the reproducer above.

**Comment by dionhaefner** [comment]
Sweet, thanks - another bug slain ⚔️
